contracts/mmr/MMRVerification.sol

```diff
@@ -8,124 +8,140 @@ pragma solidity ^0.7.0;
         }
 
-        // Check target peak found
-        if(targetPeak == bytes32(0)) {
-            return false;
-        }
+        // Calculate our leaf's mountain peak hash
+        bytes32 mountainHash = calculatePeakRoot(
+            numLeftPeaks, leafNodeHash, leafPos, targetPeakIndex, proofItems
+        );
 
-        // Find the path climbing down
-        uint256[] memory path = new uint256[](siblings.length + 1);
-        uint256 left;
-        uint256 right;
-        uint8 height = uint8(siblings.length) + 1;
-        while (height > 0) {
-            // Record the current cursor and climb down
-            path[--height] = cursor;
-            if (cursor == index) {
-                // On the leaf node. Stop climbing down
-                break;
-            } else {
-                // On the parent node. Go left or right
-                (left, right) = getChildren(cursor);
-                cursor = index > left ? right : left;
-                continue;
-            }
+        // All right peaks are rolled up into one hash. If there are any, bag them.
+        bytes32 bagger = mountainHash;
+        if(targetPeakIndex > leafPos) {
+            bagger = keccak256(abi.encodePacked(proofItems[proofItems.length-1], bagger));
         }
 
-        // // Calculate the summit hash climbing up again
-        bytes32 node;
-        while (height < path.length) {
-            // Move cursor
-            cursor = path[height];
-            if (height == 0) {
-                // cursor is on the leaf
-                node = value;
-            } else if (cursor - 1 == path[height - 1]) {
-                // cursor is on a parent and a sibling is on the left
-                node = hashParent(siblings[height - 1], node);
-            } else {
-                // cursor is on a parent and a sibling is on the right
-                node = hashParent(node, siblings[height - 1]);
-            }
-            // Climb up
-            height++;
+        // Bag left peaks one-by-one
+        for(uint i = numLeftPeaks; i > 0; i--) {
+            bagger = keccak256(abi.encodePacked(bagger, proofItems[i-1]));
         }
 
-        // Computed hash value of the summit should equal to the target peak hash
-        if(node != targetPeak) {
-            return false;
-        }
-        return true;
+        return bagger == root;
     }
 
-  // TODO: Fix so it works with for variable input array lengths
-    function combineBaggedPeaks(bytes32[] memory peakBagging)
-        public
-        pure
-        returns (bytes32 root)
-    {
-        bytes32 temp;
-        for (uint256 i = peakBagging.length - 1; i > 0; i--) {
-            if (i == peakBagging.length - 1) {
-                temp = keccak256(
-                    abi.encodePacked(peakBagging[i], peakBagging[i - 1])
-                );
+    /**
+     * @dev Calculate a leaf's mountain peak based on it's hash, it's position,
+     *      the mountain peak's position, and the proof contents.
+     */
```

Comment:
Could you define somewhere in this code what a `position`, `index`, `height` and `size` means?

Reply:
Added definitions and an example 7-leaf MMR in https://github.com/Snowfork/verification-solidity/pull/8/commits/4e52d1690593bc18efa81a121a90bff009c35da3